### DISCUSSION TITLE
Audio lag on Windows 10

### DISCUSSION THREAD

**Discussion by postacik** [timschneeb/GalaxyBudsClient#9] [General]
Hi,
This is not a bug or question about this app but since you've mastered the bt communications with the buds, I thought you might help.

Buds+ has no audio lag while watching movies or Youtube videos on my phone (Galaxy S8) but on Windows 10 v1809 (Dell Precision 3520 laptop) it lags about 100 ms which is very annoying.

Is there a definitive solution for this problem? Thanks :)

**Comment by timschneeb** [comment]
Hmm, I'm not that sure how to fix this issue. 

Apparently, the YouTube app on Android compensates for this lag by delaying the video to match the audio latency.
Also, Samsung devices come with their own [scalable proprietary audio codec](https://www.samsung.com/in/support/mobile-devices/what-is-scalable-codec/) which usually fix this lag on your (Samsung) phone. If they'd port this optimized codec as an audio driver to Windows then this should definitely get rid of the problem.

I also tried to add some registry-keys to alter the scanning (and connection) behavior of the Bluetooth audio profile from here:
<https://docs.microsoft.com/en-us/windows-hardware/drivers/bluetooth/bluetooth-registry-entries#scanning-parameterization-settings>
No luck though.

You could also try to force Windows to use the AAC codec instead of SBC, but apparently the driver of your Bluetooth adapter has to support that codec. Windows don't seem to have a generic AAC Bluetooth codec installed by default.

**Comment by postacik** [comment]
There are many Google results for this problem but none of them is useful for the latency problem with TWS headphones on Windows.

As far as I can understand you also have the exact same problem and have not found a proper solution.

I hope someone else will come up with a solution after reading this issue. :)

**Comment by superbonaci** [comment]
On Android, either audio is delayed, or video is advanced automatically, for most media, so you can watch YouTube or use local players without asynchronization, at least so little that is not noticeable. It depends also on the earbuds, on some chinese ones the delay is not corrected and can be up to half a second. Also, this is only for a/v media. If you play games (example [Brawl Stars](https://play.google.com/store/apps/details?id=com.supercell.brawlstars)), there's no way they can advance the image for you, so there will be always a minor delay, no matter how good the headset is.
No idea if Android reads the delay from device specifications or uses the same value for all.

Some earbuds have special low latency modes for gaming, like [Razer Hammerhead True Wireless Earbuds](https://www.razer.com/mobile-earphones/razer-hammerhead-true-wireless-earbuds/RZ12-02970100-R3U1), which promises as low as 60ms. There are others like [Realme Buds Air](https://www.gearbest.com/community/Ci944S) | [Realme Buds Air Neo](https://technave.com/gadget/The-Undefeated-realme-Buds-Air-Neo-with-super-low-latency-of-119-2ms-19024.html), but no comparison with wired headphones for competitive gaming.

About Windows, latest update 2004 can display battery percentage for most earbuds on bluetooth settings, which version 1909 didn't have. For Samsung Galaxy Buds it does, for Apple Airpods Pro doesn't. Will take a look later on Windows if compensates the delay or still doesn't have such feature.

On Linux both games and youtube have a noticeable delay. But hey, bluetooth integration on Linux is still on its infancy. There are lots of other things to get working first.

- [Windows 10 version 2004 now supports battery level for our AirDots natively!](https://www.reddit.com/r/airdots/comments/he4b0u/windows_10_version_2004_now_supports_battery/)
- [Future versions of Windows 10 will show battery level of Bluetooth accessories](https://mspoweruser.com/future-versions-of-windows-10-will-show-battery-level-of-bluetooth-accessories/)
- [Is there any way can get the bluetooth battery level information?](https://docs.microsoft.com/en-us/answers/questions/32945/is-there-any-way-can-get-the-bluetooth-battery-lev.html)

**Comment by postacik** [comment]
So are the following statements correct?

Bluetooth has a natural amount of latency because of its communication speed and/or used codecs.

The latency is compensated by video player apps on some operating systems like Android to sync video to audio.

Windows does not have this compensation built in to the OS, yet.

-------------

I could not understand why you mentioned about displaying battery percentage on Windows? Is it something related to audio delay?

------------

On Windows, I can sync audio of movies by using the audio delay function on VLC Media Player (hotkeys k and j). However it's still annoying to do it every time I open a video.

But I could not find a way to do it while playing Youtube videos on Chrome. It would be helpful if someone knows how.

**Reply by Ryan1981** [reply]
Have you tried http://ac3filter.sourceforge.net/ac3filter_eng.html, it has delay functions for every speaker.

**Comment by superbonaci** [comment]
@postacik Try [YouTube Audio/Video Sync](https://chrome.google.com/webstore/detail/youtube-audiovideo-sync/mknmhikmjljhpccebpnplhicmcfjkgbk) or [Audio Sync for Youtube](https://chrome.google.com/webstore/detail/audio-sync-for-youtube/ccjlbaeijnlkjegkkooopdpaboalefgi). I can't even use them because Samsung buds keeps disconnecting for me on Windows 10 2004.

As last resource you can use VLC (which only plays youtube 720p) or [SMPlayer from SVN + mpv 0.7.0 + youtube-dl](https://blog.smplayer.info/how-to-play-a-1080p-video-from-youtube-with-smplayer/), which will allow you to choose specific quality or use best (even 4k 60FPS, but require powerful hardware). Paste the URL, maybe playlists even work.

So don't update yet to W10 2004, keep it on 1909. You can use gpedit.msc to block updates. I can't rollback because it's been more than 10 days.

**Comment by postacik** [comment]
@superbonaci , thank you for the links. [YouTube Audio/Video Sync ](https://chrome.google.com/webstore/detail/youtube-audiovideo-sync/mknmhikmjljhpccebpnplhicmcfjkgbk/related) seems to work quite well although it stutters for some time at the beginning of the video.

**Comment by timschneeb** [comment]
> So don't update yet to W10 2004, keep it on 1909. You can use gpedit.msc to block updates. I can't rollback because it's been more than 10 days.

@superbonaci Surprisingly, I used W10 build 2004 for over one month now (even back then when it was only available for insider-program members) without any Bluetooth issues at all.
Maybe build 2004 breaks compatibility with your Bluetooth adapter/driver or something like that.

**Comment by timschneeb** [comment]
> No idea if Android reads the delay from device specifications or uses the same value for all.

In the AVDTP Bluetooth protocol, the slave/sink device (Buds) requests a delay report (after the device has connected) and sends the result back to the host/source device in order to achieve synchronized audio/video playback (the same method the YouTube app *probably* uses).

You can find more detailed information about this in the [AVDTP protocol specification](https://www.docdroid.net/v3x4GlA/avdtp-spec-v13-pdf#page=42) under `6.17 Delay Reporting` (scroll a few lines up).

On my Samsung S8 (running an AOSP Android 10 ROM without Samsung's scalable codec) I have massive 380ms latency! (according to these reports)

**Comment by superbonaci** [comment]
> 
> 
> @superbonaci , thank you for the links. [YouTube Audio/Video Sync ](https://chrome.google.com/webstore/detail/youtube-audiovideo-sync/mknmhikmjljhpccebpnplhicmcfjkgbk/related) seems to work quite well although it stutters for some time at the beginning of the video.

If Google Chrome addons don't work you can try with Firefox addons, but there are not many options.

If somebody has macOS, would like to know if it auto-corrects automatically like on Android or it's just like Windows.

**Comment by superbonaci** [comment]
> 
> 
> > So don't update yet to W10 2004, keep it on 1909. You can use gpedit.msc to block updates. I can't rollback because it's been more than 10 days.
> 
> @superbonaci Surprisingly, I used W10 build 2004 for over one month now (even back then when it was only available for insider-program members) without any Bluetooth issues at all.
> Maybe build 2004 breaks compatibility with your Bluetooth adapter/driver or something like that.

Tried again after reboot just in case, and happens with any bluetooth earphones. When I click on connect it quickly connects and disconnects. Can't even remove paired devices (W10 2004, worked fine on 1909):

When hitting Remove device:
![Samsung Galaxy Buds on android devices - Removing device](https://i.imgur.com/0s51c5K.png)

![Samsung Galaxy Buds on android devices - Remove failed](https://i.imgur.com/lgE11ZV.png)

When hitting Connect:
![Samsung Galaxy Buds on android devices - That didn't work. Make sure your Bluetooth device is still discoverable, then try again](https://i.imgur.com/A5jRruH.png)

Is there any way to view debug/error mensages on W10? Tried with Event Viewer but doesn't display anything.

**Comment by timschneeb** [comment]
Maybe you could try to remove/uninstall it with the Device manager instead.
You can also view all events related to the device directly in the device manager here: `(Device manager > Galaxy Buds > Properties > Events)`.

**Comment by superbonaci** [comment]
> 
> 
> Maybe you could try to remove/uninstall it with the Device manager instead.
> You can also view all events related to the device directly in the device manager here: `(Device manager > Galaxy Buds > Properties > Events)`.

Galaxy Buds don't appear under Device Manager -> PC -> Bluetooth. Tried for the Airpods Pro (which do appear) and same thing, quickly disconnects and reconnects but only 2 events appear from time ago: `Device configured (bth.inf)` and `Device started`.

**Comment by superbonaci** [comment]
@ThePBone do you have mac hardware to test if macOS displays battery percentage natively for various headsets?

**Comment by timschneeb** [comment]
@superbonaci Nope, It don't own any Apple devices

**Comment by superbonaci** [comment]
@ThePBone after some tests on a virtual machine, macOS only displays battery percentage for Apple headsets. Say Airpods Pro and other Apple products. For other brands doesn't display anything although pairing and sound work.

[How to check AirPods Battery on iPhone, Watch, Mac, and Android](https://appletoolbox.com/check-airpods-battery-life/)

[WH 1000 XM3 battery level on mac](https://www.reddit.com/r/sony/comments/exk774/wh_1000_xm3_battery_level_on_mac/)

**Comment by WY-WY-W** [comment]
How force Windows to use the AAC codec instead of SBC???

**Comment by superbonaci** [comment]
> 
> 
> How force Windows to use the AAC codec instead of SBC???

According to Reddit, W10 does not support AAC.

[Does Windows 10/8.1 support AAC bluetooth codec and how do I know what it's using?](https://www.reddit.com/r/windows/comments/bi0jfr/does_windows_1081_support_aac_bluetooth_codec_and/)

**Comment by postacik** [comment]
I've recently upgraded the BIOS firmware and drivers on my Windows 10 laptop and my company sometimes push OS updates. I don't know which one had an effect but the audio lag with my Galaxy Buds+ on Windows 10 (version 1809) reduced dramatically which means it's OK while watching videos with VLC Player or on Youtube. When I play the "Audio Video Sync Test" video on the following page, it feels like 50 ms lag. It was about 100 before. 

https://www.youtube.com/watch?v=ucZl6vQ_8Uo

**Comment by superbonaci** [comment]
@postacik maybe it was Windows update 20H2. You can check it with Run -> winver

**Comment by gadgetsan** [comment]
i don't think it was 20H2, i've also updated to it and i still have the latency issue (Buds+ though). BIOS update is worth looking into. I don't know if the issue is related but i can't make my microphone work on Windows 10

**Comment by superbonaci** [comment]
@gadgetsan probably you have the latency issues on Windows with Buds+ because it's not using the Samsung proprietary codec, which does when you connect them to a Samsung phone. The codec is the SSC (Samsung Scalable codecs).

**Comment by timschneeb** [comment]
[Meta] Converted this issue to a discussion thread (new beta feature) because it fits better

**Comment by Ryan1981** [comment]
First of all, thank you for this great piece of software for my galaxy buds!

Second, what I consider the holy grail for these buds is the following: 

https://superuser.com/questions/930251/possible-to-use-bluetooth-headset-in-stereo-mode-with-mic

My current workaround is to use a seperate microphone in my onboard sound card which works fine but is less neat than using the mic on the buds.

Would be awesome if you could unofficially get this to work, suddenly everyone would have a nice wireless gaming headset!

**Reply by timschneeb** [reply]
Unfortunately, this won't ever happen for any Samsung Galaxy Buds model because they don't support the aptX codec. This is mentioned in the thread you linked:
> Your headset needs to support aptX (https://www.aptx.com/products). This is a limitation of the bluetooth A2DP profile. Without aptX, it will not be able to receive hifi stereo sound one way while sending microphone audio the other way.

Even then, aptX headphones can still have this limitation.

**Reply by Ryan1981** [reply]
Alright, thanks for the reply, was worth asking since I feel it is technically possible (bandwidth large enough) but nobody bothered to implement it. Anyway keep up the great work! 👍

**Comment by gustavorios2929** [comment]
It's possible to add game mode to GalaxyBudsClient like in wear app?

Don't know if my test can help on this matter but heres go nothing:

I've connected my pc to my galaxy phone using steam link with my buds live connected to the phone. 
It had noticeable latency.
In wear app I enabled game mode and no noticeable latency. Everything working perfectly (except that I cant use my phone while doing this =[ )
